prover/crates/lib/prover_job_processor/src/job_runner.rs

```diff
@@ -0,0 +1,69 @@
+use tokio::task::JoinHandle;
+
+use crate::{
+    task_wiring::{JobPickerTask, JobSaverTask, Task, WorkerPool},
+    BackoffAndCancellable, Executor, JobPicker, JobSaver,
+};
+
+/// It's preferred to have a minimal amount of jobs in flight at any given time.
+/// This ensures that memory usage is minimized, in case of failures, a small amount of jobs is lost and
+/// components can apply back pressure to each other in case of misconfiguration.
+const CHANNEL_SIZE: usize = 1;
+
+/// The "framework" wrapper that runs the entire machinery.
+/// Job Runner is responsible for tying together tasks (picker, executor, saver) and starting them.
+#[derive(Debug)]
+pub struct JobRunner<E, P, S>
+where
+    E: Executor,
+    P: JobPicker,
+    S: JobSaver,
```

Comment:
Do you need types of all three? Perhaps you could just generically parametrize it by data types, not processors?

Something like 

```
pub struct JobRunner<Input, Output, Metadata>
where
    <.. >
    { executor: Executor<Input, Output, Metadata>,
    picker: JobPicker<Input, Metadata>,
   saver: JobSaver<OPutput, Metadata>
}
```